src/enc_effects.c

```diff
@@ -0,0 +1,339 @@
+
+inline BOOL EncEffects_Galactic(u32 effect)
+{
+    return (effect == ENCEFF_GALACTIC_GRUNT)
+            || (effect == ENCEFF_GALACTIC_CMDR)
+            || (effect == ENCEFF_GALACTIC_CYRUS);
+}
+
+static u32 EncEffects_GetEffectPair (const UnkStruct_ov6_02240D5C *battleParams)
+{
+    u32 battleType = battleParams->unk_00;
+
+    if (battleType & BATTLE_TYPE_TRAINER) {
+        u32 trainerEffect = EncEffects_TrainerClassEffect(battleParams->unk_28[1].unk_01);
+
+        if (battleType & BATTLE_TYPE_FRONTIER) {
+            if (trainerEffect == ENCEFF_FRONTIER_BRAIN) {
+                return trainerEffect;
+            }
+
+            if (battleType & BATTLE_TYPE_DOUBLES) {
+                return ENCEFF_DOUBLE_BATTLE;
+            }
+
+            return ENCEFF_FRONTIER;
+        }
+
+        if (EncEffects_Galactic(trainerEffect)) {
+            return trainerEffect;
+        }
+
+        // There is a bug here that is never realized in as-released Platinum:
+        // When any single trainer with special encounter effects is challenged
+        // to a double battle, that trainer will lose all of those special
+        // encounter effects and be treated as any normal double battle.
+        if (battleType & BATTLE_TYPE_DOUBLES) {
+            if (trainerEffect == ENCEFF_LEADER_VOLKNER) {
+                return ENCEFF_DOUBLE_LEADER;
+            }
+
+            return ENCEFF_DOUBLE_BATTLE;
+        }
+
+        if (battleType & BATTLE_TYPE_LINK) {
+            return ENCEFF_LINK_BATTLE;
+        }
+
+        return trainerEffect;
+    }
+
+    u32 pokemonEffect = EncEffects_WildPokemonEffect(battleParams->unk_04[1], battleParams->unk_134);
+
+    if (pokemonEffect < ENCEFF_NORMAL_WILD) {
+        return pokemonEffect;
+    }
+
+    if (battleType & BATTLE_TYPE_DOUBLES) {
+        return ENCEFF_DOUBLE_WILD;
+    }
+
+    return pokemonEffect;
+}
+
+static u32 EncEffects_CutInEffectForPair (u32 effectID, const UnkStruct_ov6_02240D5C *battleParams)
+{
+    GF_ASSERT(effectID < ENCEFF_MAX);
+
+    // If the cut-in effect specifies, determine what effect to use
+    // based on the encounter locale.
+    if (sEncEffectsTable[effectID].cutInEffect == ENCEFF_CUTIN_USE_LOCAL) {
+        return CutInEffects_ForBattle(battleParams);
+    }
+    
+    return sEncEffectsTable[effectID].cutInEffect;
+}
+
+static u32 EncEffects_BGMForPair (u32 effectID, const UnkStruct_ov6_02240D5C *battleParams)
+{
+    GF_ASSERT(effectID < ENCEFF_MAX);
+
+    return sEncEffectsTable[effectID].sdatBGMusic;
+}
+
+u32 EncEffects_CutInEffect (const UnkStruct_ov6_02240D5C *battleParams)
+{
+    return EncEffects_CutInEffectForPair(EncEffects_GetEffectPair(battleParams), battleParams);
+}
+
+u32 EncEffects_BGM (const UnkStruct_ov6_02240D5C *battleParams)
+{
+    return EncEffects_BGMForPair(EncEffects_GetEffectPair(battleParams), battleParams);
+}
+
+static u32 EncEffects_TrainerClassEffect (u32 trainerClass)
```

Comment:
I tried to simplify both this function and the one below for wild Pokemon in order to get rid of the `result` declaration at the top and instead issue a `return` for each `case`. That, unfortunately, didn't match, so here we are.